f/common_logic/data_validation.py

```diff
@@ -0,0 +1,152 @@
+import csv
+import json
+import logging
+from pathlib import Path
+
+import pandas as pd
```

Comment:
I notice `openpyxl` in the test requirements (and not in the requirements to run these functions).  Why's it needed there?

Or to take the opposite approach, could we use just `openpyxl` instead of all of pandas to implement these functions?

Reasons to prefer pandas:
- very widely known
- reads easily
- it might (?? I'm guessing) provide a good fallback if openpyxl is not able to be installed.
- it is more in-line with a suggestion of mine down below to consider Fiona for spatial formats

Reasons to prefer openpyxl
- I am guessing pandas installs it already anyway
- more lightweight
- more directly solves your intent
- less likely to end up with version conflicts with other application code, i.e. our scripts are likely to install pandas themselves

Reply:
So, the reason we need `openpyxl` in the test requirements is because we use pandas `df.to_excel` which uses `openpyxl` as an engine for writing, and as it turns out, that does need to be installed separately. So without having this in `requirements.txt`, the one test which leverages that method will fail with a ModuleNotFoundError.

https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.to_excel.html

(Turns out you can also use `xlsxwriter` as an engine, but that requires installation too.)

I will add an inline comment to clarify this in requirements.

I'll also make a note to clarify the current approach, and possibilities of revisiting (and also using `fiona`) in the future.

Reply:
You are right. Fixing this.